Applications/Spire/Source/Ui/TextBox.cpp

```diff
@@ -263,85 +398,60 @@ TextBox::TextBox(QString current, QWidget* parent)
-    reinterpret_cast<std::intptr_t>(this)));
-  m_line_edit->setFrame(false);
-  m_line_edit->setTextMargins(-2, 0, -4, 0);
-  m_line_edit->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
-  auto& highlight = m_highlight->get();
-  m_line_edit->setCursorPosition(highlight.m_end);
-  m_line_edit->setSelection(highlight.m_start, get_size(highlight));
-  m_text_validator = new TextValidator(m_current, this);
-  m_line_edit->setValidator(m_text_validator);
-  m_line_edit->installEventFilter(this);
-  setCursor(m_line_edit->cursor());
-  setFocusPolicy(m_line_edit->focusPolicy());
-  m_box = new PlaceholderBox(m_line_edit);
-  auto layout = new QHBoxLayout(this);
-  layout->setContentsMargins(0, 0, 0, 0);
-  layout->addWidget(m_box);
-  setFocusProxy(m_box);
-  proxy_style(*this, *m_box);
-  add_pseudo_element(*this, Placeholder());
+      m_line_edit(nullptr) {
   m_style_connection = connect_style_signal(*this, [=] { on_style(); });
-  m_placeholder_style_connection =
-    connect_style_signal(*this, Placeholder(), [=] { on_style(); });
   set_style(*this, DEFAULT_STYLE());
-  connect(m_line_edit, &QLineEdit::editingFinished, this,
-    &TextBox::on_editing_finished);
-  connect(m_line_edit, &QLineEdit::textEdited, this, &TextBox::on_text_edited);
   m_current_connection = m_current->connect_update_signal(
     [=] (const auto& value) { on_current(value); });
-  connect(m_line_edit, &QLineEdit::cursorPositionChanged, this,
-    std::bind_front(&TextBox::on_cursor_position, this));
-  connect(m_line_edit, &QLineEdit::selectionChanged, this,
-    std::bind_front(&TextBox::on_selection, this));
-  m_highlight->connect_update_signal(
-    std::bind_front(&TextBox::on_highlight, this));
+  m_submission->connect_update_signal([=] (const auto& value) {
+    on_submission(value);
+  });
 }
 
 const std::shared_ptr<TextModel>& TextBox::get_current() const {
   return m_current;
 }
 
 const QString& TextBox::get_submission() const {
-  return m_submission;
+  return m_submission->get();
 }
 
 const std::shared_ptr<HighlightModel>& TextBox::get_highlight() const {
   return m_highlight;
 }
 
 void TextBox::set_placeholder(const QString& placeholder) {
-  m_box->set_placeholder(placeholder);
-  update_placeholder_text();
+  m_placeholder = placeholder;
+  if(m_line_edit) {
+    m_line_edit->set_placeholder(m_placeholder);
+  }
 }
 
 bool TextBox::is_read_only() const {
-  return m_line_edit->isReadOnly();
+  return m_read_only->get();
 }
 
 void TextBox::set_read_only(bool read_only) {
-  if(m_line_edit->isReadOnly() == read_only) {
+  if(read_only == m_read_only->get()) {
+    return;
+  }
+  m_read_only->set(read_only);
+  if(!m_line_edit && m_read_only->get()) {
+    match(*this, ReadOnly());
     return;
+  } else if(!m_line_edit) {
+    initialize_line_edit();
   }
-  m_line_edit->setReadOnly(read_only);
-  m_line_edit->setCursorPosition(0);
   setCursor(m_line_edit->cursor());
-  if(read_only) {
+  if(m_read_only->get()) {
```

Comment:
Similarly there's a lot of context switching here. It really is as simple as if the component is read-only, match the read only selector, otherwise unmatch it. Avoid intertwining different non-local aspects of a model.